### DISCUSSION TITLE
How can I add language support for the editor?

### DISCUSSION THREAD

**Discussion by NoahGWood** [eclipse-theia/theia#10086] [Q&A] [answered]
I'm giving up and asking here because I just do not grok this

I tried building a language server in VSCode following Microsoft's tutorial, can't get the extension to run in Theia.
I tried generating language support with the yeoman generator, doesn't work.
I followed the official website's example, both for a plugin and for an extension. Doesn't work.
I finally found a tutorial online and thought maybe my quest was over, but alas, this also did not work.

I even used the yeoman to create a front-end extension so I could verify the extension was actually being built and the "hello world" functioned fine, but I did not have language support.

My final coup-de-grace I poured over error messages, it looks like I needed vscode-textmate, so I added that. And then all my builds failed, I tried a bunch of versions at random (latest, 1.0.0, 3.5.0), etc. even after deleting all of these packages (completely) I can't even get it to build anymore. I'm beyond my wits end with this dependency hell.

Any suggestions on where I can even begin to restart this? I'm on day 5 trying to just implement the most basic comment highlighting functionality and back at square one again.

**Comment by msujew** [comment]
Hi @NoahGWood, thank you for your question. You should've come here earlier, we're always eager to help :)

In short: The easiest way should be build a language server based extension for vscode. Usually those should work out-of-the-box in Theia as soon as they successfully run in vscode. Even more complicated language server extensions like the one for TypeScript/JavaScript and Java are suppported by Theia. Would you mind sharing a reproducible example for your vscode extension that wouldn't run in Theia?

For basic syntax based highlighting, you should also be able to follow the examples in the documentation using the `LanguageGrammarDefinitionContribution`. However adding any more language features through Theia's extensions system isn't supported anymore, as the LSP support is only available for vscode/theia plugins. (FYI there is a clear distinction between Theia plugins and extensions, [see here](https://theia-ide.org/docs/authoring_extensions/).

> I finally found a tutorial online and thought maybe my quest was over

Depending on how old they are, most tutorials regarding Theia might be outdated. So I would treat with care there.

**Reply by NoahGWood** [reply]
> For basic syntax based highlighting, you should also be able to follow the examples in the documentation using the LanguageGrammarDefinitionContribution

Yeah, I tried that as a last resort thinking I should be able to at least get syntax highlighting, but I've not been able to get that to work either.

> Would you mind sharing a reproducible language server in vscode..

I was trying to use the lsp-sample from https://github.com/microsoft/vscode-extension-samples it works fine in VSCode, but I haven't been able to import it as a plugin or an extension.

**Comment by NoahGWood** [comment, marked as answer]
Finally got something to work, I guess I needed to 

1. Use VSCode to create and test extension.
2. Use VCSE to package the extension
3. Save the package in ~/.theia/extension/
4. Include: `
    "@theia/plugin-dev": "1.17.2",
    "@theia/plugin-ext": "1.17.2",
    "@theia/plugin-ext-vscode": "1.17.2"` in my package.json
